presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestRecordingHiveMetastore.java

```diff
@@ -0,0 +1,579 @@
+                    OptionalLong.of(8))));
+    private static final HivePrivilegeInfo PRIVILEGE_INFO = new HivePrivilegeInfo(HivePrivilege.SELECT, true);
+
+    @Test
+    public void testRecordingHiveMetastore()
+            throws IOException
+    {
+        HiveClientConfig recordingHiveClientConfig = new HiveClientConfig()
+                .setRecordingPath(File.createTempFile("recording_test", "json").getAbsolutePath())
+                .setRecordingDuration(new Duration(10, TimeUnit.MINUTES));
+
+        RecordingHiveMetastore recordingHiveMetastore = new RecordingHiveMetastore(new TestingHiveMetastore(), recordingHiveClientConfig);
+        validateMetadata(recordingHiveMetastore);
+        recordingHiveMetastore.dropDatabase("other_database");
+        recordingHiveMetastore.writeRecording();
+
+        HiveClientConfig replayingHiveClientConfig = recordingHiveClientConfig
+                .setReplay(true);
+
+        recordingHiveMetastore = new RecordingHiveMetastore(new UnimplementedHiveMetastore(), replayingHiveClientConfig);
+        recordingHiveMetastore.loadRecording();
+        validateMetadata(recordingHiveMetastore);
+    }
+
+    private void validateMetadata(ExtendedHiveMetastore hiveMetastore)
+    {
+        assertEquals(hiveMetastore.getDatabase("database"), Optional.of(DATABASE));
+        assertEquals(hiveMetastore.getAllDatabases(), ImmutableList.of("database"));
+        assertEquals(hiveMetastore.getTable("database", "table"), Optional.of(TABLE));
+        assertEquals(hiveMetastore.getSupportedColumnStatistics(createVarcharType(123)), ImmutableSet.of(MIN_VALUE, MAX_VALUE));
+        assertEquals(hiveMetastore.getTableStatistics("database", "table"), PARTITION_STATISTICS);
+        assertEquals(hiveMetastore.getPartitionStatistics("database", "table", ImmutableSet.of("value")), ImmutableMap.of("value", PARTITION_STATISTICS));
+        assertEquals(hiveMetastore.getAllTables("database"), Optional.of(ImmutableList.of("table")));
+        assertEquals(hiveMetastore.getAllViews("database"), Optional.empty());
+        assertEquals(hiveMetastore.getPartition("database", "table", ImmutableList.of("value")), Optional.of(PARTITION));
+        assertEquals(hiveMetastore.getPartitionNames("database", "table"), Optional.of(ImmutableList.of("value")));
+        assertEquals(hiveMetastore.getPartitionNamesByParts("database", "table", ImmutableList.of("value")), Optional.of(ImmutableList.of("value")));
+        assertEquals(hiveMetastore.getPartitionsByNames("database", "table", ImmutableList.of("value")), ImmutableMap.of("value", Optional.of(PARTITION)));
+        assertEquals(hiveMetastore.getRoles("user"), ImmutableSet.of("role1", "role2"));
+        assertEquals(hiveMetastore.getDatabasePrivileges("user", "database"), ImmutableSet.of(PRIVILEGE_INFO));
+        assertEquals(hiveMetastore.getTablePrivileges("user", "database", "table"), ImmutableSet.of(PRIVILEGE_INFO));
+    }
+
+    private static class TestingHiveMetastore
```

Comment:
Why not extend `UnimplementedHiveMetastore`? Then you don't need to implement all the methods.